File: example/tibble_etf_returns.f90
program tibble_etf_returns
use r_mod, only: dp, print_tibble, read_csv_tibble_real, r_tibble_real_t, &
   tibble_ncol, tibble_nrow, tibble_real_log_returns, tibble_real_stats
implicit none

character(len=512) :: csv_path
type(r_tibble_real_t) :: csv_tbl, price_tbl, return_tbl, stats_tbl
logical :: exists

csv_path = "asset_class_etf_prices.csv"
if (command_argument_count() >= 1) call get_command_argument(1, csv_path)
inquire(file=trim(csv_path), exist=exists)
if (.not. exists) then
   write(*, '(a)') "ETF return example skipped: CSV file not found"
   write(*, '(a)') "Pass its path after --, for example:"
   write(*, '(a)') "fpm run --example tibble_etf_returns -- asset_class_etf_prices.csv"
   stop
end if

csv_tbl = read_csv_tibble_real(trim(csv_path), max_cols=4, index_col="Date")
if (tibble_ncol(csv_tbl) < 1 .or. tibble_nrow(csv_tbl) < 2) &
   error stop "tibble_etf_returns: expected dated price columns"
price_tbl = csv_tbl
return_tbl = tibble_real_log_returns(price_tbl, scale=100.0_dp)
stats_tbl = tibble_real_stats(return_tbl)

write(*, '(/, a)') repeat("=", 72)
write(*, '(a)') "r_mod example: ETF log returns in a real-only tibble"
write(*, '(a, /)') repeat("=", 72)
write(*, '(a, 1x, a)') "Price file:", trim(csv_path)
write(*, '(a, i0)') "Price observations: ", tibble_nrow(price_tbl)
write(*, '(a, i0)') "Complete return observations: ", tibble_nrow(return_tbl)
write(*, '(a, i0)') "Assets: ", tibble_ncol(return_tbl)
write(*, '(/, a)') "First five scaled log returns:"
call print_tibble(return_tbl, 5)

write(*, '(/, a)') "Return statistics (returns are multiplied by 100):"
call print_tibble(stats_tbl, integer_row_labels=[character(len=1) :: "n"], &
   decimal_places=4, row_numbers=.false.)
end program tibble_etf_returns

File: example/tibble_etf_returns.R
args <- commandArgs(trailingOnly = TRUE)
csv_path <- if (length(args) >= 1) args[[1]] else "asset_class_etf_prices.csv"

if (!file.exists(csv_path)) {
  cat("ETF return example skipped: CSV file not found\n")
  cat("Pass its path as the first argument, for example:\n")
  cat("Rscript example/tibble_etf_returns.R asset_class_etf_prices.csv\n")
  quit(status = 0)
}

if (!requireNamespace("tibble", quietly = TRUE)) {
  stop("This example requires the tibble package.")
}

csv_tbl <- tibble::as_tibble(read.csv(csv_path, check.names = FALSE))
if (!("Date" %in% names(csv_tbl))) {
  stop("tibble_etf_returns: expected a Date column")
}

asset_names <- c("SPY", "EFA", "EEM", "EMB")
if (nrow(csv_tbl) < 2) {
  stop("tibble_etf_returns: expected dated price columns")
}

price_tbl <- csv_tbl[, asset_names]
price_matrix <- as.matrix(price_tbl)
return_matrix <- 100 * log(
  price_matrix[-1, , drop = FALSE] /
    price_matrix[-nrow(price_matrix), , drop = FALSE]
)
keep <- rowSums(is.finite(return_matrix)) == ncol(return_matrix)
return_matrix <- return_matrix[keep, , drop = FALSE]
colnames(return_matrix) <- asset_names
return_dates <- csv_tbl$Date[-1]
return_dates <- return_dates[keep]

return_tbl <- tibble::as_tibble(return_matrix)
return_tbl <- tibble::add_column(
  return_tbl,
  Date = return_dates,
  .before = 1
)

stats_matrix <- rbind(
  n = 1.0 * colSums(is.finite(return_matrix)),
  mean = colMeans(return_matrix),
  sd = apply(return_matrix, 2, sd),
  minimum = apply(return_matrix, 2, min),
  maximum = apply(return_matrix, 2, max)
)
colnames(stats_matrix) <- asset_names
rownames(stats_matrix) <- c("n", "mean", "sd", "minimum", "maximum")
stats_tbl <- tibble::as_tibble(stats_matrix, rownames = "statistic")

cat("\n", strrep("=", 72), "\n", sep = "")
cat("R example: ETF log returns in a tibble\n")
cat(strrep("=", 72), "\n\n", sep = "")
cat("Price file:", csv_path, "\n")
cat("Price observations:", nrow(price_tbl), "\n")
cat("Complete return observations:", nrow(return_tbl), "\n")
cat("Assets:", length(asset_names), "\n")
cat("\nFirst five scaled log returns:\n")
print(head(return_tbl, 5))

cat("\nReturn statistics (returns are multiplied by 100):\n")
cat("Statistic rows:", nrow(stats_tbl), "\n")
print(stats_tbl)
